sys/sleepq.c

```diff
@@ -209,21 +214,37 @@ void sleepq_wait(void *wchan, const void *waitpt) {
       td->td_state = TDS_SLEEPING;
       sched_switch();
     }
+    /* TDF_SLPINTR flag is set after wakeup only if sleep was aborted. */
+    if (td->td_flags & TDF_SLPINTR) {
+      td->td_flags &= ~TDF_SLPINTR;
+      wakeup = SQ_ABORT;
+    }
   }
+
+  return wakeup;
 }
 
 /* Remove a thread from the sleep queue and resume it. */
-static void sq_wakeup(thread_t *td, sleepq_chain_t *sc, sleepq_t *sq) {
+static bool sq_wakeup(thread_t *td, sleepq_chain_t *sc, sleepq_t *sq,
+                      sq_wakeup_t wakeup) {
+  /* Do not try to abort thread's sleep if it's not prepared for that. */
+  if (!(td->td_flags & TDF_SLPINTR) && (wakeup == SQ_ABORT))
```

Comment:
Is it okay to check `td_flags` without `td_spin`?

Reply:
I believe it's ok. I added a comment that tries to justify it.